### PR TITLE
:recycle: :white_check_mark: Refactor and add tests for tuple

### PR DESCRIPTION
**Pull request by bdeane-intel**
In this commit:
- testing tuple type_traits
- compile-time tests as well as runtime tests
- value category tests
- added tests for apply and fold

### CODE REVIEW

File: test/cib/tuple.cpp [hunk 1]
```diff
@@ -21,37 +24,198 @@ struct J {};
 struct K {};
 struct L {};
 
-TEST_CASE("make empty cib::tuple", "[tuple]") {
-    auto const t = cib::make_tuple();
+TEMPLATE_TEST_CASE("constructability", "[tuple]",
```
Comment (mjcaisse-intel) [inline test/cib/tuple.cpp:27]: 👏  woot! Look at all of these tests.

**Review by mjcaisse-intel — COMMENTED** [review]
Still working on the review.

**Review by mjcaisse-intel — APPROVED** [review]
Looks really great Ben!

File: include/cib/tuple.hpp [hunk 2]
```diff
@@ -9,65 +13,63 @@
...
+    }
 };
 
 template <typename T, std::size_t Index, typename... IndexTs>
 struct tuple_element
     : type_indexed_element<tuple_element<T, Index, IndexTs...>, IndexTs>... {
     CIB_CONSTEXPR static auto index = Index;
     using value_type = T;
-    T value{};
+    value_type value;
```
[the remark this answers is not in the document]
Comment (bdeane-intel) [reply include/cib/tuple.hpp]: Yes - I've changed this back so that cib::tuple is not trivially constructible, same as std::tuple (and std::pair)

**Pull request merged**